Fix `mfc.bat run` for MS Windows

henryleberre opened the pull request:
## Description

Small changes in the toolchain to support `./mfc.sh run` on Windows. Note: I was unable to build MFC (because of a local environment issue) so I did testing without a binary present.

sbryngelson commented:
two things
> Note: I was unable to build MFC (because of a local environment issue) so I did testing without a binary present.

what does this mean exactly?

Can you also explain what your changes to the mako files?

sbryngelson commented:
Here's the error on the CPU job for benchmarking:
```console
 .:++=----------====+*= ==..:%..... | --targets pre_process, simulation, and post_process
  .:-=++++===--==+=-+=   +.  :=     | 
  +#=::::::::=%=. -+:    =+   *:    | ----------------------------------------------------------
 .*=-=*=..    :=+*+:      -...--    | $ ./mfc.sh (build, run, test, clean, count, packer) --help


 Benchmarking pre_process, simulation, and post_process (build/benchmarks/b298):
  
   1/4: 5eq_rk3_weno3_hllc @ benchmarks/5eq_rk3_weno3_hllc/case.py
    
     > Log:     build/benchmarks/b298/5eq_rk3_weno3_hllc.out
     > Summary: build/benchmarks/b298/5eq_rk3_weno3_hllc.yaml
     $ ./mfc.sh run /storage/scratch1/6/sbryngelson3/runners/actions-runner-2/_work/MFC/MFC/pr/benchmarks/5eq_rk3_weno3_hllc/case.py --case-optimization --targets pre_process simulation post_process --output-summary /storage/scratch1/6/sbryngelson3/runners/actions-runner-2/_work/MFC/MFC/pr/build/benchmarks/b298/5eq_rk3_weno3_hllc.yaml -c phoenix -n 4 -- 1

 

Error: Failed to load YAML from "/storage/scratch1/6/sbryngelson3/runners/actions-runner-2/_work/MFC/MFC/pr/build/benchmarks/b298/5eq_rk3_weno3_hllc.yaml": mapping values are not allowed here
  in "/storage/scratch1/6/sbryngelson3/runners/actions-runner-2/_work/MFC/MFC/pr/build/benchmarks/b298/5eq_rk3_weno3_hllc.yaml", line 23, column 17

./mfc.sh: line 51: 159489 Terminated              python3 "$(pwd)/toolchain/main.py" "$@"

mfc: ERROR > mfc.py finished with a 143 exit code.
mfc: (venv) Exiting the Python virtual environment.
```

ChrisZYJ commented:
I tested it on Windows and it works great after some small changes to default.mako and test/case.py. I'll submit these changes after this PR

henryleberre commented:
> two things
> 
> > Note: I was unable to build MFC (because of a local environment issue) so I did testing without a binary present.
> 
> what does this mean exactly?
> 
> Can you also explain what your changes to the mako files?

I couldn't get CMake to detect the Intel compilers so I couldn't build MFC. I do not understand why this is the case, it used to work. As a result, I wasn't able to `./mfc.bat build`.

This PR fixes two things:
- `mfc.bat`: It was referencing a python file that has since moved.
- `mfc.bat run`: The issue lies with the Mako files. When we switched to Mako, I forgot to take Windows into consideration. Therefore, running `mfc.bat run` would generate a (Bash) unix shell script to execute MFC. I fixed it so that the default Mako file generates a Windows Batch file if on Windows that behaves about the same as the Unix one.

ChrisZYJ commented:
Yeah the Intel compilers are really frustrating (WSL is so much better). I've added detailed steps in the docs explaining how I set it up. Also, with some tweaks, the code now runs great natively on Windows and can pass the test suites. The only issue is that pre-processing only works on a single core due to some write permission problems, but simulation works fine with MPI.

henryleberre commented on docs/documentation/getting-started.md:
```diff
@@ -59,29 +60,52 @@ On Windows, you can either use Intel Compilers with the standard Microsoft toolc
+- [Intel® oneAPI Base Toolkit](https://www.intel.com/content/www/us/en/developer/tools/oneapi/base-toolkit-download.html)
+- [Intel® oneAPI HPC Toolkit](https://www.intel.com/content/www/us/en/developer/tools/oneapi/hpc-toolkit-download.html)
+- [Strawberry Perl](https://strawberryperl.com/) (Install and add `C:\strawberry\perl\bin\perl.exe` or your installation path to your [PATH](https://www.architectryan.com/2018/03/17/add-to-the-path-on-windows-10/))
+Please note that Visual Studio must be installed first, and the oneAPI Toolkits need to be configured with the installed Visual Studio, even if you plan to use a different IDE.
```
@ChrisZYJ Could you explain what you mean by this?
> the oneAPI Toolkits need to be configured with the installed Visual Studio

This might be what I was missing!

ChrisZYJ replied:
It's this step:
![image](https://github.com/user-attachments/assets/e00a77a3-7706-4d82-94bf-507e06acdf31)

I heard it wouldn't work without this, but not very sure

henryleberre replied:
I see what you mean, that makes sense. I also did this.

ChrisZYJ replied:
For me I had to upgrade both Visual Studio and oneAPI to their 2024 versions before CMake managed to find the Fortran compiler

henryleberre replied:
That gives me hope! I am now updating my VS 2022.

henryleberre commented:
Thanks @ChrisZYJ for adding documentation and fixing some bugs! I read somewhere that some of the intel errors can be caused by Windows Defender analyzing build files (the process takes a file handle and prohibits others from accessing it). This might be the cause for the need to run `mfc.bat build` twice for each target - but I am not sure.

ChrisZYJ commented:
> Thanks @ChrisZYJ for adding documentation and fixing some bugs! I read somewhere that some of the intel errors can be caused by Windows Defender analyzing build files (the process takes a file handle and prohibits others from accessing it). This might be the cause for the need to run `mfc.bat build` twice for each target - but I am not sure. Back when I first implemented Windows support, at least three or four builds of each target were required (iirc).

One rebuild for each target is bad enough - I wonder how long 3 or 4 rebuilds took... It's also so much slower to build on Windows than Linux (especially when initializing the compilers)

henryleberre commented:
I still see some strange behavior like this:
```console
C:\Users\Henry Le Berre\Desktop\MFC>mfc.bat build -t pre_process -j 8
Comparing files C:\USERS\HENRY LE BERRE\DESKTOP\MFC\BUILD\requirements.txt and C:\USERS\HENRY LE BERRE\DESKTOP\MFC\TOOLCHAIN\REQUIREMENTS.TXT
FC: no differences encountered

      .=++*:          -+*+=.        | Henry Le Berre@DESKTOP-ET4J2MI [Windows]
     :+   -*-        ==   =* .      | ----------------------------------------
   :*+      ==      ++    .+-       |
  :*##-.....:*+   .#%+++=--+=:::.   | --jobs 8
  -=-++-======#=--**+++==+*++=::-:. | --no-mpi --no-gpu --no-debug --no-gcov --no-unified
 .:++=----------====+*= ==..:%..... | --targets pre_process
  .:-=++++===--==+=-+=   +.  :=     |
  +#=::::::::=%=. -+:    =+   *:    | ----------------------------------------------------------
 .*=-=*=..    :=+*+:      -...--    | $ ./mfc.sh (build, run, test, clean, count, packer) --help

 Build | syscheck and pre_process | Generic Build

 Generating case.fpp.
   Writing a (new) custom case.fpp file.
 $ cmake -DMFC_SYSCHECK=ON -Wno-dev --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=C:\Users\Henry Le Berre\Desktop\MFC\build\install\5ce829ed14;C:\Users\Henry Le Berre\Desktop\MFC\build\install\dependencies -DCMAKE_FIND_ROOT_PATH=C:\Users\Henry Le Berre\Desktop\MFC\build\install\5ce829ed14;C:\Users\Henry Le Berre\Desktop\MFC\build\install\dependencies -DCMAKE_INSTALL_PREFIX=C:\Users\Henry Le Berre\Desktop\MFC\build\install\5ce829ed14 -DCMAKE_Fortran_MODULE_DIRECTORY=build/install/dependencies/include/hipfort/amdgcn -DMFC_MPI=OFF -DMFC_OpenACC=OFF -DMFC_GCov=OFF -DMFC_Unified=OFF -S C:\Users\Henry Le Berre\Desktop\MFC\ -B C:\Users\Henry Le Berre\Desktop\MFC\build\staging\5ce829ed14

Not searching for unused variables given on the command line.
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.40.33812.0
-- The CXX compiler identification is MSVC 19.40.33812.0
-- The Fortran compiler identification is Intel 2021.0.0.20240602
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Determine Intel Fortran Compiler Implicit Link Path
-- Determine Intel Fortran Compiler Implicit Link Path - done
-- Check for working Fortran compiler: C:/Program Files (x86)/Intel/oneAPI/compiler/2024.2/bin/ifort.exe - skipped
-- Performing Test SUPPORTS_MARCH_NATIVE
-- Performing Test SUPPORTS_MARCH_NATIVE - Failed
-- Performing Test SUPPORTS_MCPU_NATIVE
-- Performing Test SUPPORTS_MCPU_NATIVE - Failed
-- IPO / LTO is NOT available
-- Configuring done (99.7s)
-- Generating done (0.0s)
-- Build files have been written to: C:/Users/Henry Le Berre/Desktop/MFC/build/staging/5ce829ed14

 Generating case.fpp.
   Writing a (new) custom case.fpp file.
 $ cmake --build C:\Users\Henry Le Berre\Desktop\MFC\build\staging\5ce829ed14 --target syscheck --parallel 8 --config Release


Microsoft Visual Studio 2022 Version 17.10.4.
Copyright (C) Microsoft Corp. All rights reserved.
Build started at 14:51...
1>------ Build started: Project: ZERO_CHECK, Configuration: Release x64 ------
1>Checking File Globs
1>1>Checking Build System
2>------ Build started: Project: syscheck, Configuration: Release x64 ------
Preprocessing (Fypp) syscheck.fpp
Compiling with Intel® Fortran Compiler Classic 2021.12.0 [Intel(R) 64]...
syscheck.fpp.f90
ifort: remark #10448: Intel(R) Fortran Compiler Classic (ifort) is now deprecated and will be discontinued late 2024. Intel recommends that customers transition now to using the LLVM-based Intel(R) Fortran Compiler (ifx) for continued Windows* and Linux* support, new language support, new language features, and optimizations. Use '/Qdiag-disable:10448' to disable this message.
Linking...
Link: executing 'C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\bin\HostX64\x64\link.exe'
Microsoft (R) Incremental Linker Version 14.40.33812.0
Copyright (C) Microsoft Corporation.  All rights reserved.
"/OUT:C:\Users\Henry Le Berre\Desktop\MFC\build\staging\5ce829ed14\Release\syscheck.exe"
/VERSION:0.0
/INCREMENTAL:NO
/MANIFEST
/MANIFESTFILE:syscheck.dir\Release\syscheck.exe.intermediate.manifest
"/MANIFESTUAC:level='asInvoker' uiAccess='false'"
/SUBSYSTEM:CONSOLE
"/IMPLIB:C:\Users\Henry Le Berre\Desktop\MFC\build\staging\5ce829ed14\Release\syscheck.lib"
user32.lib
/machine:x64
syscheck.dir\Release\syscheck.fpp.obj
Embedding manifest...
Microsoft (R) Manifest Tool
Copyright (c) Microsoft Corporation.
All rights reserved.

Build log written to  "file://C:/Users/Henry%20Le%20Berre/Desktop/MFC/build/staging/5ce829ed14/syscheck.dir/Release/BuildLog.htm"
syscheck - 0 error(s), 0 warning(s)
========== Build: 2 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 14:51 and took 04,152 seconds ==========

 $ cmake --install C:\Users\Henry Le Berre\Desktop\MFC\build\staging\5ce829ed14

-- Install configuration: "Release"
-- Installing: C:/Users/Henry Le Berre/Desktop/MFC/build/install/5ce829ed14/bin/syscheck.exe

 Generating case.fpp.
   Writing a (new) custom case.fpp file.
 $ cmake -DMFC_PRE_PROCESS=ON -Wno-dev --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=C:\Users\Henry Le Berre\Desktop\MFC\build\install\6cb7d810e7;C:\Users\Henry Le Berre\Desktop\MFC\build\install\dependencies -DCMAKE_FIND_ROOT_PATH=C:\Users\Henry Le Berre\Desktop\MFC\build\install\6cb7d810e7;C:\Users\Henry Le Berre\Desktop\MFC\build\install\dependencies -DCMAKE_INSTALL_PREFIX=C:\Users\Henry Le Berre\Desktop\MFC\build\install\6cb7d810e7 -DCMAKE_Fortran_MODULE_DIRECTORY=build/install/dependencies/include/hipfort/amdgcn -DMFC_MPI=OFF -DMFC_OpenACC=OFF -DMFC_GCov=OFF -DMFC_Unified=OFF -S C:\Users\Henry Le Berre\Desktop\MFC\ -B C:\Users\Henry Le Berre\Desktop\MFC\build\staging\6cb7d810e7

Not searching for unused variables given on the command line.
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.40.33812.0
-- The CXX compiler identification is MSVC 19.40.33812.0
-- The Fortran compiler identification is unknown
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:15 (project):
  No CMAKE_Fortran_COMPILER could be found.



-- Configuring incomplete, errors occurred!


Error: Failed to configure the pre_process target.


[mfc.bat] Failed to run MFC.
```

It found Intel Fortran for `syscheck` but not for `pre_process`.. I don't think it's worth it for me to delve deeper into this, a clean reinstall of windows would probably solve the problem. I am satisfied with you getting it working.

@sbryngelson I think this PR is ready for review/merge.
